ms-oidc/src/main/java/net/trajano/ms/oidc/internal/OpenIdConnectResource.java

```diff
@@ -109,28 +136,56 @@ public OAuthTokenResponse callback(@QueryParam("code") final String code,
             throw OAuthTokenResponse.badRequest(ErrorCodes.INVALID_REQUEST, "Invalid issuer_id");
         }
 
+        final ServerState serverState = nonceCache.get(jwtState, ServerState.class);
+        if (serverState == null) {
+            throw OAuthTokenResponse.badRequest(ErrorCodes.INVALID_REQUEST, "Invalid state");
+        }
+
         final URI redirectUri = UriBuilder.fromUri(serviceConfiguration.getRedirectUri()).path(issuerId).build();
         final Form form = new Form();
         form.param("redirect_uri", redirectUri.toASCIIString());
-        form.param("grant_type", "authorization_code");
+        form.param("grant_type", GrantTypes.AUTHORIZATION_CODE);
         form.param("code", code);
         final OpenIdConfiguration openIdConfiguration = issuerConfig.getOpenIdConfiguration();
         final IdTokenResponse openIdToken = client.target(openIdConfiguration.getTokenEndpoint())
             .request(MediaType.APPLICATION_JSON)
             .header(HttpHeaders.AUTHORIZATION, issuerConfig.buildAuthorization())
             .post(Entity.form(form), IdTokenResponse.class);
 
-        final JwtClaims jwtClaims = cryptoOps.toClaimsSet(openIdToken.getIdToken(), openIdConfiguration.getHttpsJwks());
+        final JwtClaims idTokenClaims = cryptoOps.toClaimsSet(openIdToken.getIdToken(), openIdConfiguration.getHttpsJwks());
+        if (!serverState.getNonce().equals(idTokenClaims.getStringClaimValue("nonce"))) {
+            throw OAuthTokenResponse.internalServerError("nonce did not match");
+        }
+
+        // The following would be an op that would've been performed
 
         final Form storeInternalForm = new Form();
         storeInternalForm.param("grant_type", GrantTypes.JWT_ASSERTION);
-        storeInternalForm.param("assertion", cryptoOps.sign(jwtClaims));
+        storeInternalForm.param("assertion", cryptoOps.sign(idTokenClaims));
         storeInternalForm.param("aud", issuerConfig.getClientId());
 
-        return client.target(authorizationEndpoint).request(MediaType.APPLICATION_JSON)
-            .header(HttpHeaders.AUTHORIZATION, HttpAuthorizationHeaders.buildBasicAuthorization(appClientId, appClientSecret))
+        // TODO the authorization endpoint should change depending on the value of "op", one end point could be authorization and another could be registration.
```

Comment:
![INFO](https://sonarsource.github.io/sonar-github/severity-info.png 'Severity: INFO') Complete the task associated to this TODO comment. [![rule](https://sonarsource.github.io/sonar-github/rule.png)](https://sonarcloud.io/coding_rules#rule_key=squid%3AS1135)